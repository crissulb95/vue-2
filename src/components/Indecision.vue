<template>
  <div class="indecision-container">
    <figure class="img-container">
      <img v-if="bgGif" :src="bgGif" alt="bg" />
    </figure>
    <div class="input-container">
      <input
        type="text"
        placeholder="Haz una pregunta"
        v-model="question"
        data-test-id="input-question"
      />
      <small>Recuerda terminar con un signo de interrogaci√≥n (?)</small>
    </div>
    <div v-if="isValidQuestion" class="question-container">
      <p>{{ question }}</p>
      <h3>{{ answer }}</h3>
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {
      question: "primer valor",
      answer: null,
      bgGif: null,
      isValidQuestion: false,
    };
  },
  methods: {
    async getAnswer() {
      try {
        this.answer = "pensando...";

        const { answer, image } = await fetch("https://yesno.wtf/api")
          .then((res) => res.json())
          .catch((err) => console.error(err));

        this.answer =
          answer === "yes" ? "Si" : answer === "maybe" ? "Tal vez" : "No!";
        this.bgGif = image;
      } catch (error) {
        this.answer = "Hubo un error en la carga desde el API";
        this.bgGif = null;
        console.log(
          "IndecisionComponent: Error en la carga desde el API",
          error
        );
      }
    },
  },
  watch: {
    question(val) {
      this.isValidQuestion = false;
      console.log({ val });

      if (!val.includes("?")) return false;

      console.log({ val });
      this.isValidQuestion = true;
      this.getAnswer();
    },
  },
};
</script>

<style scoped>
.indecision-container {
  display: block;
  width: 50%;
  flex: 0 0 50%;
  position: relative;
  padding: 150px 20px;
  margin: 0;
}

.indecision-container .img-container {
  height: 100%;
  left: 50%;
  margin: 0;
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%);
  width: 100%;
  z-index: 0;
}
.indecision-container .img-container::after {
  background-color: rgba(0, 0, 0, 0.3);
  content: "";
  height: 100%;
  left: 50%;
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%);
  width: 100%;
  z-index: 1;
}
.indecision-container .img-container img {
  height: 100%;
  width: 100%;
  position: absolute;
  top: 50%;
  left: 50%;
  object-fit: cover;
  object-position: center;
  transform: translate(-50%, -50%);
}

.input-container {
  position: relative;
  z-index: 2;
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: center;
}

.input-container > input {
  padding: 15px;
  flex: 0 0 50%;
  border-radius: 5px;
  margin-bottom: 10px;
}
.input-container > small {
  flex: 0 0 100%;
}

.question-container {
  display: flex;
  flex: 0 0 100%;
  position: relative;
  z-index: 2;
  flex-wrap: wrap;
  margin-top: 30px;
}
.question-container p,
.question-container h3 {
  display: block;
  width: 100%;
  flex: 0 0 100%;
  min-height: 20px;
}
</style>
