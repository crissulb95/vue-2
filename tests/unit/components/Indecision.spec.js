import Indecision from "@/components/Indecision";
import { shallowMount } from "@vue/test-utils";

describe("Indecision Component", () => {
  let wrapper;
  let clgSpy;

  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: "yes",
          forced: false,
          image: "https://yesno.wtf/assets/yes/2.gif",
        }),
    })
  );

  beforeEach(() => {
    wrapper = shallowMount(Indecision);
    clgSpy = jest.spyOn(console, "log");
    jest.clearAllMocks();
  });

  test("Debería corresponder con el snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("No debe llamar a la función getAnswer", async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    const input = wrapper.find("[data-test-id='input-question']");

    await input.setValue("Test text");

    expect(clgSpy).toHaveBeenCalled();
    expect(getAnswerSpy).not.toHaveBeenCalled();
    //expect(clgSpy).toHaveBeenCalledTimes(1);
  });
  test('Al escribir "?" debe disparar el llamado a la función getAnswer', async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    const input = wrapper.find("[data-test-id='input-question']");

    await input.setValue("Test text2?");

    expect(clgSpy).toHaveBeenCalledTimes(2);
    expect(getAnswerSpy).toHaveBeenCalledTimes(1);
  });
  test("Pruebas en getAnswer()", async () => {
    await wrapper.vm.getAnswer();

    const gif = wrapper.find(".img-container img");
    //console.log(gif.exists());
    expect(gif.exists()).toBeTruthy();

    console.log("answer: ", wrapper.vm.answer);
    console.log("Background GIF: ", wrapper.vm.bgGif);
    expect(wrapper.vm.answer).toBe("Si");
    expect(wrapper.vm.bgGif).toBe("https://yesno.wtf/assets/yes/2.gif");
  });
  test("Pruebas en getAnswer() deben fallar", async () => {
    fetch.mockImplementationOnce(() => Promise.reject("API is unavailable"));
    await wrapper.vm.getAnswer();

    const gif = wrapper.find(".img-container img");
    expect(gif.exists()).toBeFalsy();
    expect(wrapper.vm.answer).toBe("Hubo un error en la carga desde el API");
  });
});
