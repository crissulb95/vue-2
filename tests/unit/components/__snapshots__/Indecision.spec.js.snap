// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component Debería corresponder con el snapshot 1`] = `
<div class="indecision-container">
  <figure class="img-container">
    <!--v-if-->
  </figure>
  <div class="input-container"><input type="text" placeholder="Haz una pregunta" data-test-id="input-question"><small>Recuerda terminar con un signo de interrogación (?)</small></div>
  <!--v-if-->
</div>
`;
